File: src/Modules/Helpdesk/Data/Sla.cs
namespace Modules.Helpdesk.Data;

[Flags]
public enum BusinessDays
{
    None = 0,
    Monday = 1,
    Tuesday = 2,
    Wednesday = 4,
    Thursday = 8,
    Friday = 16,
    Saturday = 32,
    Sunday = 64,
    Weekdays = Monday | Tuesday | Wednesday | Thursday | Friday,
}

public sealed class BusinessHoursCalendar
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;
    public string TimeZoneId { get; set; } = "UTC";
    public BusinessDays WorkingDays { get; set; } = BusinessDays.Weekdays;
    public TimeOnly StartTime { get; set; }
    public TimeOnly EndTime { get; set; }
    public DateTimeOffset CreatedAt { get; set; }
}

/// <summary>
/// One rule in an ordered list: the conditions a ticket must meet, and the targets it then gets.
///
/// <para>
/// Every condition is optional and null means "any", so a policy with none is the catch-all. The
/// first active policy in <see cref="SortOrder"/> whose conditions all match is the one that
/// applies — the model Zendesk and Jira Service Management use, and the reason order is a column an
/// administrator sets rather than an accident of creation time.
/// </para>
/// </summary>
public sealed class SlaPolicy
{
    public Guid Id { get; set; }
    public string Name { get; set; } = string.Empty;

    /// <summary>Lower is evaluated first. Ties fall back to <see cref="CreatedAt"/> for determinism.</summary>
    public int SortOrder { get; set; }

    public TicketPriority? Priority { get; set; }

    /// <summary>Incidents and service requests usually carry different targets; null matches both.</summary>
    public TicketType? TicketType { get; set; }

    /// <summary>
    /// Matched by id rather than by the old free-text name, so renaming a category in Settings
    /// cannot silently detach the SLA that was written for it.
    /// </summary>
    public Guid? CategoryId { get; set; }
    public TicketCategory? Category { get; set; }

    public int ResponseTargetMinutes { get; set; }
    public int ResolutionTargetMinutes { get; set; }
    public int WarningPercent { get; set; } = 80;
    public Guid CalendarId { get; set; }
    public BusinessHoursCalendar Calendar { get; set; } = null!;
    public bool IsActive { get; set; } = true;
    public DateTimeOffset CreatedAt { get; set; }
}

/// <summary>
/// One ticket's clock.
///
/// <para>
/// The targets are <b>copied here</b> when the policy attaches rather than read through it. Editing
/// a policy would otherwise re-target every ticket already running against it — tighten a target at
/// lunchtime and work that was on track is retrospectively breached, with notifications firing and
/// no record that anything moved. Every real SLA engine snapshots for this reason, and
/// <see cref="PolicyId"/> is kept only to say which rule was applied.
/// </para>
/// </summary>
public sealed class TicketSla
{
    public Guid Id { get; set; }
    public Guid TicketId { get; set; }
    public Ticket Ticket { get; set; } = null!;

    /// <summary>Which policy attached. Provenance only — nothing is measured through it.</summary>
    public Guid PolicyId { get; set; }
    public SlaPolicy Policy { get; set; } = null!;

    public int ResponseTargetMinutes { get; set; }
    public int ResolutionTargetMinutes { get; set; }
    public int WarningPercent { get; set; }
    public Guid CalendarId { get; set; }
    public BusinessHoursCalendar Calendar { get; set; } = null!;

    public DateTimeOffset StartedAt { get; set; }
    public DateTimeOffset? ActiveSince { get; set; }
    public double AccumulatedBusinessSeconds { get; set; }
    public double? ResponseBusinessSeconds { get; set; }
    public DateTimeOffset? ResponseCompletedAt { get; set; }
    public DateTimeOffset? ResolutionCompletedAt { get; set; }
    public bool ResponseWarningRaised { get; set; }
    public bool ResolutionWarningRaised { get; set; }
    public bool ResponseBreached { get; set; }
    public bool ResolutionBreached { get; set; }
}

File: web/src/api/sla.ts
import { apiRequest } from './client'
import type { TicketPriority, TicketType } from './helpdesk'

/**
 * Which days a calendar counts, as the server's [Flags] enum. Sunday is 64 rather than 0 because the
 * flags are a bit set, not a day index.
 */
export const businessDays = [
  { value: 1, label: 'Mon' },
  { value: 2, label: 'Tue' },
  { value: 4, label: 'Wed' },
  { value: 8, label: 'Thu' },
  { value: 16, label: 'Fri' },
  { value: 32, label: 'Sat' },
  { value: 64, label: 'Sun' },
] as const

export type BusinessHoursCalendar = {
  id: string
  name: string
  timeZoneId: string
  /** A bit set over `businessDays`. */
  workingDays: number
  startTime: string
  endTime: string
  /** How many policies measure against it. One with any cannot be deleted. */
  policyCount: number
}

/**
 * One rule in an ordered list. Every condition is optional and null means "any", so a policy with
 * none is the catch-all; the first active policy in order whose conditions all hold is the one a new
 * ticket gets.
 */
export type SlaPolicy = {
  id: string
  name: string
  sortOrder: number
  priority: TicketPriority | null
  ticketType: TicketType | null
  categoryId: string | null
  categoryName: string | null
  responseTargetMinutes: number
  resolutionTargetMinutes: number
  warningPercent: number
  calendarId: string
  calendarName: string
  isActive: boolean
  /** Tickets that have run against it. Any at all means it can be deactivated but not deleted. */
  ticketCount: number
}

export type SavePolicyInput = {
  name: string
  responseTargetMinutes: number
  resolutionTargetMinutes: number
  warningPercent: number
  calendarId: string
  priority?: TicketPriority | null
  ticketType?: TicketType | null
  categoryId?: string | null
  sortOrder?: number
  isActive?: boolean
}

export type SaveCalendarInput = {
  name: string
  timeZoneId: string
  workingDays: number
  startTime: string
  endTime: string
}

export const slaApi = {
  listPolicies: () => apiRequest<SlaPolicy[]>('/api/sla/policies'),
  createPolicy: (input: SavePolicyInput) =>
    apiRequest<SlaPolicy>('/api/sla/policies', { method: 'POST', body: JSON.stringify(input) }),
  updatePolicy: (id: string, input: SavePolicyInput) =>
    apiRequest<SlaPolicy>(`/api/sla/policies/${id}`, { method: 'PUT', body: JSON.stringify(input) }),
  deletePolicy: (id: string) => apiRequest<void>(`/api/sla/policies/${id}`, { method: 'DELETE' }),
  /** Answers with the whole list in its new order, so the screen never has to guess. */
  reorderPolicies: (policyIds: string[]) =>
    apiRequest<SlaPolicy[]>('/api/sla/policies/order', { method: 'POST', body: JSON.stringify({ policyIds }) }),

  listCalendars: () => apiRequest<BusinessHoursCalendar[]>('/api/sla/calendars'),
  createCalendar: (input: SaveCalendarInput) =>
    apiRequest<BusinessHoursCalendar>('/api/sla/calendars', { method: 'POST', body: JSON.stringify(input) }),
  deleteCalendar: (id: string) => apiRequest<void>(`/api/sla/calendars/${id}`, { method: 'DELETE' }),
}

/** What a policy's conditions read as, for the list. */
export function describePolicyConditions(policy: SlaPolicy): string {
  const parts = [
    policy.priority ?? 'Any priority',
    policy.ticketType === 'ServiceRequest' ? 'Service requests'
      : policy.ticketType === 'Incident' ? 'Incidents'
        : 'Any kind',
  ]
  if (policy.categoryName) parts.push(policy.categoryName)
  return parts.join(' · ')
}

/** Minutes as something readable: targets are entered in minutes but read in hours and days. */
export function describeMinutes(minutes: number): string {
  if (minutes < 60) return `${minutes}m`
  if (minutes % (60 * 24) === 0) return `${minutes / (60 * 24)}d`
  if (minutes % 60 === 0) return `${minutes / 60}h`
  return `${Math.floor(minutes / 60)}h ${minutes % 60}m`
}
